marker_api/src/context.rs

```diff
@@ -308,48 +308,61 @@ impl<'ast> MarkerContext<'ast> {
+    pub resolve_ty_ids: extern "C" fn(&'ast MarkerContextData, path: ffi::FfiStr<'_>) -> ffi::FfiSlice<'ast, TyDefId>,
 
     // Internal utility
-    pub expr_ty: extern "C" fn(&'ast (), ExprId) -> SemTyKind<'ast>,
-    pub span: extern "C" fn(&'ast (), SpanId) -> &'ast Span<'ast>,
-    pub span_snippet: extern "C" fn(&'ast (), &Span<'ast>) -> ffi::FfiOption<ffi::FfiStr<'ast>>,
-    pub span_source: extern "C" fn(&'ast (), &Span<'_>) -> SpanSource<'ast>,
-    pub span_pos_to_file_loc: extern "C" fn(&'ast (), &FileInfo<'ast>, SpanPos) -> ffi::FfiOption<FilePos<'ast>>,
-    pub span_expn_info: extern "C" fn(&'ast (), ExpnId) -> ffi::FfiOption<&'ast ExpnInfo<'ast>>,
-    pub symbol_str: extern "C" fn(&'ast (), SymbolId) -> ffi::FfiStr<'ast>,
-    pub resolve_method_target: extern "C" fn(&'ast (), ExprId) -> ItemId,
+    pub expr_ty: extern "C" fn(&'ast MarkerContextData, ExprId) -> SemTyKind<'ast>,
+    pub span: extern "C" fn(&'ast MarkerContextData, SpanId) -> &'ast Span<'ast>,
+    pub span_snippet: extern "C" fn(&'ast MarkerContextData, &Span<'ast>) -> ffi::FfiOption<ffi::FfiStr<'ast>>,
+    pub span_source: extern "C" fn(&'ast MarkerContextData, &Span<'_>) -> SpanSource<'ast>,
+    pub span_pos_to_file_loc:
+        extern "C" fn(&'ast MarkerContextData, &FileInfo<'ast>, SpanPos) -> ffi::FfiOption<FilePos<'ast>>,
+    pub span_expn_info: extern "C" fn(&'ast MarkerContextData, ExpnId) -> ffi::FfiOption<&'ast ExpnInfo<'ast>>,
+    pub symbol_str: extern "C" fn(&'ast MarkerContextData, SymbolId) -> ffi::FfiStr<'ast>,
+    pub resolve_method_target: extern "C" fn(&'ast MarkerContextData, ExprId) -> ItemId,
 }
 
 impl<'ast> MarkerContextCallbacks<'ast> {
     fn call_emit_diagnostic<'a>(&self, diag: &'a Diagnostic<'a, 'ast>) {
-        (self.emit_diag)(self.driver_context, diag);
+        (self.emit_diag)(self.data, diag);
     }
 
     fn call_expr_ty(&self, expr: ExprId) -> SemTyKind<'ast> {
-        (self.expr_ty)(self.driver_context, expr)
+        (self.expr_ty)(self.data, expr)
     }
     fn call_span(&self, span_id: SpanId) -> &'ast Span<'ast> {
-        (self.span)(self.driver_context, span_id)
+        (self.span)(self.data, span_id)
     }
     fn call_symbol_str(&self, sym: SymbolId) -> &'ast str {
-        (self.symbol_str)(self.driver_context, sym).get()
+        (self.symbol_str)(self.data, sym).get()
     }
     pub fn resolve_method_target(&self, expr: ExprId) -> ItemId {
-        (self.resolve_method_target)(self.driver_context, expr)
+        (self.resolve_method_target)(self.data, expr)
     }
 }
+
+/// This type is used by [`MarkerContextCallbacks`] as the first argument to every
+/// function. For more information, see the documentation of the `data` field
+/// or from `marker_adapter::context`.
+///
+/// This type should never be constructed and is only meant as a pointer
+/// casting target.
+#[repr(C)]
+#[cfg_attr(feature = "driver-api", visibility::make(pub))]
+struct MarkerContextData {
```

Comment:
Not sure that having this struct removes the need for callbacks to be unsafe themselves. Because you can still trigger UB from safe Rust, although now only within this crate like this:
```rust
(self.callbacks.resolve_ty_ids)(&MarkerContextData { _data: 0 }, path.into())
```
Anyway, that's not too important to postpone this PR. We can refactor unsafe any time. It still works this way, so that should be fine

Reply:
It doesn't remove the potential for UB, but makes it way less likely, since the type is private to the module. So, this is a slight improvement. We can add unsafe later :)